control_plane/attachment_service/src/service.rs

```diff
@@ -1266,6 +1270,299 @@ impl Service {
+                            return Err(ApiError::Conflict(
+                                "Cannot split, currently mid-split".to_string(),
+                            ));
+                        }
+                    }
+                }
+                if policy.is_none() {
+                    policy = Some(shard.policy.clone());
+                }
+                if shard_ident.is_none() {
+                    shard_ident = Some(shard.shard);
+                }
+
+                if tenant_shard_id.shard_count == ShardCount(split_req.new_shard_count) {
+                    tracing::warn!(
+                        "Tenant shard {} already has shard count {}",
+                        tenant_shard_id,
+                        split_req.new_shard_count
+                    );
+                    continue;
+                }
+
+                let node_id =
+                    shard
+                        .intent
+                        .attached
+                        .ok_or(ApiError::BadRequest(anyhow::anyhow!(
+                            "Cannot split a tenant that is not attached"
+                        )))?;
+
+                let node = pageservers
+                    .get(&node_id)
+                    .expect("Pageservers may not be deleted while referenced");
+
+                // TODO: if any reconciliation is currently in progress for this shard, wait for it.
+
+                targets.push(SplitTarget {
+                    parent_id: *tenant_shard_id,
+                    node: node.clone(),
+                    child_ids: tenant_shard_id.split(ShardCount(split_req.new_shard_count)),
+                });
+            }
+
+            if targets.is_empty() {
+                if children_found.len() == split_req.new_shard_count as usize {
+                    return Ok(TenantShardSplitResponse {
+                        new_shards: children_found,
+                    });
+                } else {
+                    // No shards found to split, and no existing children found: the
+                    // tenant doesn't exist at all.
+                    return Err(ApiError::NotFound(
+                        anyhow::anyhow!("Tenant {} not found", tenant_id).into(),
+                    ));
+                }
+            }
+
+            (old_shard_count, targets, locked.compute_hook.clone())
+        };
+
+        // unwrap safety: we would have returned above if we didn't find at least one shard to split
+        let old_shard_count = old_shard_count.unwrap();
+        let shard_ident = shard_ident.unwrap();
+        let policy = policy.unwrap();
+
+        // Before creating any new child shards in memory or on the pageservers, persist them: this
+        // enables us to ensure that we will always be able to clean up if something goes wrong.  This also
+        // acts as the protection against two concurrent attempts to split: one of them will get a database
+        // error trying to insert the child shards.
```

Comment:
We're not holding the `self.inner.write()` lock anymore, so, at this point, another caller could call could run from start to finish.
Then we would run again and overwrite their state.

I see you're using the `UniqueViolation` to detect that.

I think the better approach would be to replace the `self.inner.write()` by locking through the database, i.e., `select for update` with `where` clause selecting the entire `TenantShardId::tenant_range(tenant_id)`, or, maybe just the entire `tenant_id`.

Also, do we actually need the `write()` lock above? I can't spot a place where the loop above is _modifying_ anything. In fact, it shouldn't, because, if it losees the race and hit `UniqueViolation` below, it would have to undo the changes, and it doesn't right now.

Reply:
Fixed the spurious write() lock in 212c0ee7c.

> We're not holding the self.inner.write() lock anymore, so, at this point, another caller could call could run from start to finish.

If another caller runs from start to finish with the same target count, they will have created the child shards, and we'll drop out on the UniqueConstraint.  There is a risk in more complicated cases though, like if the tenant is deleted after we drop.  Building the pre-checks into `Persistence::begin_shard_split` would make sense, or it might be simpler to put an async lock inside TenantState, so that for these kinds of cases one can exclude all other operations on the tenant.

I've made this a bit safer by passing the shard count into begin_shard_split, and having that function error out if during its transaction it sees that the number of parent shards is not as expected, and added a FIXME in the function to note that we should pull more of the validation part into the database transaction to make this path sounder: b9f004613